ex2_services/Services/source/scheduler/scheduler.c

```diff
@@ -7,685 +19,169 @@
-                    xSemaphoreGive(scheduleSemaphore);
-                    vPortFree(cmds);
-                    vPortFree(sorted_cmds);
-                    status = CALLOC_ERROR;
-                    memcpy(&gs_cmds->data[STATUS_BYTE], &status, sizeof(int8_t));
-                    return SATR_ERROR;
-                }
-                scheduled_commands_unix_t *updated_cmds = (scheduled_commands_unix_t *)pvPortMalloc(
-                    num_existing_cmds * sizeof(scheduled_commands_unix_t));
-                memset(updated_cmds, 0, num_existing_cmds * sizeof(scheduled_commands_unix_t));
-                if (num_existing_cmds > 0 && updated_cmds == NULL) {
-                    sys_log(ERROR, "pvPortMalloc for updated_cmds failed in DELETE_SCHEDULE, out of memory");
-                    red_close(fout);
-                    xSemaphoreGive(scheduleSemaphore);
-                    vPortFree(cmds);
-                    vPortFree(sorted_cmds);
-                    vPortFree(existing_cmds);
-                    status = CALLOC_ERROR;
-                    memcpy(&gs_cmds->data[STATUS_BYTE], &status, sizeof(int8_t));
-                    return SATR_ERROR;
-                }
-                // read file
-                red_lseek(fout, 0, RED_SEEK_SET);
-                int32_t f_read = red_read(fout, existing_cmds, (uint32_t)scheduler_stat.st_size);
-                if (f_read < 0) {
-                    sys_log(ERROR, "error %d from red_read() in DELETE_SCHEDULE for file: '%s'", (int)red_errno,
-                            fileName1);
-                    red_close(fout);
-                    xSemaphoreGive(scheduleSemaphore);
-                    vPortFree(cmds);
-                    vPortFree(sorted_cmds);
-                    vPortFree(existing_cmds);
-                    status = (int)red_errno;
-                    memcpy(&gs_cmds->data[STATUS_BYTE], &status, sizeof(int8_t));
-                    return SATR_ERROR;
+            xSemaphoreGive(SchedLock);
+
+            // Queue a message to the scheduler daemon to wake it up.
+            int ctx = 2;
+            if (xQueueSendToBack(SchedulerNotificationQueue, &ctx, 0) != pdTRUE)
+                sys_log(NOTICE, "xQueueSend failed");
+        }
+    } break;
+
+    case GET_SCHEDULE: {
+        sys_log(DEBUG, "Get Schedule request received");
+        // Open schedule file. Note that it's OK if the file doesn't exist.
+        if (xSemaphoreTake(SchedLock, (TickType_t) SCHED_SEM_WAIT) == pdTRUE) {
+            int32_t fd = red_open(ScheduleFile, RED_O_RDONLY);
+            if (fd >= 0) {
+                // get file size through file stats
+                REDSTAT fstat;
+                int err = red_fstat(fd, &fstat);
+                if (err < 0) {
+                    sys_log(NOTICE, "red_fstat error: %d", red_errno);
+                    packet->data[OUT_DATA_BYTE] = red_errno;
+                    rc = SCHED_ERR_IO;
+                    packet->length = 3;
```

Comment:
Should you close the file here?

Reply:
There is a "red_close(fd)" on line 156

Reply:
GitHub doesn't seem to handle these large diffs very nicely, I'll pull down the changes and review them locally instead